### DISCUSSION TITLE
Performance: How to make prisma perform a single SQL query with joins instead of multiple queries

### DISCUSSION THREAD

**Discussion by theodiablo** [prisma/prisma#12715] [Q&A]
Hi there,

When I make a query with the prisma client which includes other tables data, prisma seems to be chaining several SQL queries instead of making a single query to the database using joins.
It is making my database timeout, as it performs many select queries for very simple operations.

Prisma Version: `3.8.1`

Here is my query:
```
      const dbTransactions = await prismaClient.transaction.findMany({
        where: { id: 1,
        select: {
          profile: true,
          currency: true,
          industry: {
            select: {
              industryImpacts: {
                select: {
                  currency: true,
                },
              },
            },
          },
        },
      });
```

My model is defined like this (I deleted many of the fields to keep it simpler):
```
model Transaction {
  id                           Int                           @id @default(autoincrement())
...
  profileId                    Int
  profile                      Profile                       @relation(fields: [profileId], references: [id])
  currencyId                   Int
  currency                     Currency                      @relation(fields: [currencyId], references: [id])
  industryId                   Int
  industry                     Industry                      @relation(fields: [industryId], references: [id])

  @@index([currencyId])
  @@index([profileId])
}

model Currency {
  id              Int               @id @default(autoincrement())
...
  transactions    Transaction[]
}

model Industry {
  id                               Int                               @id @default(autoincrement())
...
  industryImpacts                  IndustryImpacts[]
}

model IndustryImpacts {
  id         Int      @id @default(autoincrement())
...
  currencyId Int
  currency   Currency @relation(fields: [currencyId], references: [id])
  industryId Int
  industry   Industry @relation(fields: [industryId], references: [id])
}
```

This is how I instantiate the `prismaClient` to be able to debug the queries:
```
export const prismaClient = new PrismaClient({
  log: [
    {
      emit: 'event',
      level: 'query',
    },
  ],
});

prismaClient.$on('query', async e => {
  console.log(`${e.query} ${e.params}`);
});
```

Here is the list of SQL queries performed by the client:
```
SELECT `Transaction`... FROM `Transaction` WHERE `Transaction`.`id` = ? [1]
SELECT `Profile`... FROM `Profile` WHERE `Profile`.`id` IN (?) [31]
SELECT `Currency`... FROM `Currency` WHERE `Currency`.`id` IN (?) [1]
SELECT `Industry`.`id` FROM `Industry` WHERE `Industry`.`id` IN (?) [43]
SELECT `IndustryImpacts`... FROM `IndustryImpacts` WHERE `IndustryImpacts`.`industryId` IN (?) [43]
```

What I would expect is some sort of intelligent joining like:
```
SELECT `Transaction`..., `Profile`..., `Currency`..., `Industry`..., `IndustryImpacts`...
FROM `Transaction` 
LEFT JOIN `Profile` ON `Profile`.`id` = `Transaction`.`profileId`
LEFT JOIN `Currency` WHERE `Currency`.`id` = `Transaction`.`currencyId`
LEFT JOIN `Industry` WHERE `Industry`.`id` = `Transaction`.`industryId`
LEFT JOIN `IndustryImpacts` WHERE `IndustryImpacts`.`industryId` = `Industry`.`id`
WHERE `Transaction`.`id` = ? [1]
```

Does someone know if there is a way that prisma makes joins like instead of calling multiple queries sequentially?

Thanks!

**Comment by janpio** [comment]
> It is making my database timeout, as it asks for many connections for very simple queries.

All these queries are executed with one and the same connection.

There is no way currently to influence the types of queries Prisma executes besides using raw queries and send plain SQL.

**Reply by kwanhs** [reply]
Is there any plan despite there is no way currently? After days of search I have seen users asking for this simple feature since Prisma 2 ?!  A simple query on my dataset took few ms on Sequelize but 30ms using your package. I could not understand why others find this bearable.

**Reply by janpio** [reply]
If you have a Prisma Client where this negatively impacts performance, of course please open a new issue and include some information so we can reproduce this @kwanhs. Most times this does not matter much, but if it does in your case that is a good argument to improve the situation.

**Reply by joar** [reply]
Aren't you throwing away any possible consistency guarantees by doing multiple SELECTs instead of one SELECT with a JOIN?

e.g.:

```sql
SELECT authorId FROM posts WHERE authorId = "foo"
-- concurrent DELETE of author where authorId = "foo";
SELECT name FROM authors WHERE authorId = "foo"
-- would return 0 rows
```
versus:
```sql
SELECT posts.authorId, authors.name FROM posts 
LEFT JOIN authors  ON  posts.authorId = authors.authorId 
WHERE posts.authorId = "foo" 
-- concurrent DELETE would not affect the SELECT's snapshot of the authors table.
```

if that's the case, how is that not a big deal?

**Reply by SabbeRubbish** [reply]
This behaviour still occurs to this date, it seems like it heavily underuses the database's potential. Feels wrong :-)

**Reply by fortizflexidao** [reply]
This is quite unsatisfactory, and it is perplexing that joins are not utilized for joining data or filtering results and not prominently mentioned in the documentation. Performing tasks in multiple queries instead of a single one results in a loss of atomicity and reduced performance. Moreover, in the case of PostgreSQL, it is crucial to have control over the order in which joins are executed, as we have observed queries with execution times reduced from 8 minutes to 800 milliseconds simply by altering the join order. Please, at the very least, emphasize this in the documentation. If it cannot be addressed due to compatibility with MongoDB, kindly make that clear as well.

**Reply by janpio** [reply]
If you have a concrete query that is to slow with Prisma, please open a bug report and provide enough information so we can reproduce it. Thanks.

**Reply by SDrinkwater** [reply]
@janpio There are several concrete examples provided in the following issues: 
- https://github.com/prisma/prisma/issues/12582
- https://github.com/prisma/prisma/issues/15616

Reproduction with tracing: https://github.com/prisma/prisma/pull/15285

Would be great if this issue could be prioritised as the current approach ends up effectively multiplying any latency present between the server and database. This results in an additional round trip for every included relation.

**Reply by MarceloPrado** [reply]
+1, just found out about this.

 I'm running a simple query using a join table, resulting in 4 different SQL queries that takes an aggregate of ~100ms. I'm running locally, with a test DB (discard any lack of indexes as cause, since my DB only has a few dozen rows).
 
 Running SQL directly results in < 30ms. Huge opportunity to improve prisma perf.
 
 I can see this getting much worse over time as my application evolves

**Reply by lukejagodzinski** [reply]
I see that I'm not alone :D. I've decided to use Prisma in my side project and as long as I was developing locally everything was fine. Then I deployed to Supabase and things got terribly slow. 500+ ms to execute "one" query which became 7 queries. I didn't even bother to check if Prisma is doing joins. I just assumed that all this hype about Prisma is justified and it has to do joins like every other ORM. But know I see I was terribly wrong. Considering this I will probably switch to writing raw SQL which makes using Prisma pointless (maybe just for schema migration). Or maybe it's time to switch to no-SQL db...

**Reply by wtachau** [reply]
+1

**Reply by Romstar** [reply]
> > It is making my database timeout, as it asks for many connections for very simple queries.
> 
> All these queries are executed with one and the same connection.
> 
> There is no way currently to influence the types of queries Prisma executes besides using raw queries and send plain SQL.

What does your statement mean exactly? Are multiple queries executed and their results are joined within Prisma's Rust based code? 

So is Prisma making multiple trips to the database to get the data via the same connection that is kept open or what exactly is going on here?

**Reply by janpio** [reply]
> What does your statement mean exactly? 

> Are multiple queries executed and their results are joined within Prisma's Rust based code?

Yes.

> So is Prisma making multiple trips to the database to get the data via the same connection that is kept open 

Yes.

**Reply by Romstar** [reply]
Thank you!

**Reply by rubek-joshi** [reply]
> Please, at the very least, emphasize this in the documentation. If it cannot be addressed due to compatibility with MongoDB, kindly make that clear as well.

@janpio I second this. This can be a deal breaker for some and is something that should definitely be conveyed to new users rather than them [finding it in an ugly way (Please have a read here)](https://codedamn.com/news/product/dont-use-prisma). I definitely think there is a good enough case to improve the current situation.

**Reply by kwanhs** [reply]
> > Please, at the very least, emphasize this in the documentation. If it cannot be addressed due to compatibility with MongoDB, kindly make that clear as well.
> 
> @janpio I second this. This can be a deal breaker for some and is something that should definitely be conveyed to new users rather than them [finding it in an ugly way (Please have a read here)](https://codedamn.com/news/product/dont-use-prisma). I definitely think there is a good enough case to improve the current situation.

It is amusing to hide behind compatibility issue. I see no points in enforcing the same SQL/commands for two different database paradigms. If MongoDB or any other NoSQL database forbids such joins, set a if condition (or fancier, class) to diverse the behaviour.

**Reply by sezanzeb** [reply]
Where exactly is it written that MongoDB compatibility is the reason for this behavior?

Anyhow, I think this would be also a deal breaker for me and I'd rather use other tools than prisma.

**Reply by gabrielgasp** [reply]
The fact that a simple query including a M-N relation like User+Roles results in Prisma making 3 queries to the database is quite shocking.

The problem here is not the time spent by the database, but the extra latency caused by a round trip for each query.

**Reply by relsunkaev** [reply]
I was trying to do a filtered relation count (recently released in GA) and it brought my whole app and db down in production because the query time went from a few ms to 20+ seconds. I spent an entire day after a rollback trying to figure out what was going on only to find out it was precisely this issue. Prisma was running a "WHERE IN" query on a 5000+ array of ids instead of doing a join.

**Reply by AnimeshRy** [reply]
We did the mistake of not checking this on a new project we started. Two of our node services use prisma and have been unbearably slow for production usecase. No point in adding more issues as prisma already has like 2k+. With all the DX hype, we did not see this clearly and now regretting it. I was really suprised to see how this was not the default behaviour from prisma.

**Reply by janpio** [reply]
If you can actually identify any slowness to the queries Prisma generates, please _do_ open an issue. We are actively working on this, and the more use cases we know about, the better we can make things work properly afterwards.

**Reply by suhjohn** [reply]
@janpio There are no slowness to the queries themselves because whatever Prisma currently generates are the simplest single line SQL selects. If this is how your team is measuring the healthiness of the feature, you better change your KPIs. 

Once you guys actually ship compiling the correct JOINs, measuring the slowness to the queries matter. Right now you don't have that feature, so obviously the performance of the SQL statement itself once it reaches the database will be trivially good. 

The issue is the # of round-trip of the SQL calls for any realistic production workload which everyone on this thread has already mentioned, which can be eliminated with JOINs. Real ORMs like ROR / Django / SQLAlchemy / Sequelize support this. Even [Drizzle](https://github.com/drizzle-team/drizzle-orm), which started 7 months ago, supports it. We'll likely migrate to Drizzle because they seem to have done everything right for SQL - better migration, better transaction, better relational queries, with the only issue being its maturity.

**Reply by rubek-joshi** [reply]
@suhjohn I wouldn't recommend Drizzle to anyone at the time of writing this. It's way too buggy for production.

The thing I'd recommend for everyone is to continue using Prisma alongside [Kysely](https://kysely.dev/) for now. You can use Prisma for easily building your models and executing the simple queries and turn towards Kysely for the queries that require JOINS. This combination works quite well when you use the [prisma-kysely](https://github.com/valtyr/prisma-kysely) library. It still isn't ideal but this is the best workaround IMO.

**Advantages of this workaround**
- You just need to add two additional libraries (kysely and prisma-kysely)
- You can still use Prisma's amazing syntax for single queries, inserts and deletes
- Prisma's generate command will automatically generate types for kysely (via prisma-kysely)
- You only need to rewrite your join queries in kysely leaving the rest as it is
- You can still make use of Prisma's migration and seeding

**Reply by janpio** [reply]
> @janpio There are no slowness to the queries themselves because whatever Prisma currently generates are the simplest single line SQL selects. If this is how your team is measuring the healthiness of the feature, you better change your KPIs.

Sorry, I was not clear: I meant the _Prisma Client queries_ as a whole. If you have concrete examples of where what Prisma does (so all the individual _SQL_ queries under the hood and the additional logic to put the data into the format you requested) is slower than it should be, then please open an issue so we have something concrete to benchmark our improvements against.

Also, because it seems to have still not been clear: We are working on this. But the underlying question in this discussion asked about the current (at the time) state, so that is what I answered. There is no need to continue trying to convince me that this is not a great state. We are working on it. You are trying to argue with someone that is agreeing with you.

**Reply by suhjohn** [reply]
Great to hear, can't wait. Fyi relevant video if you haven't seen: https://www.youtube.com/watch?v=jqhHXe746Ns. Might be good to clear the air with him too

**Reply by nastynaz** [reply]
@janpio I find it absolutely insane that an ‘ORM’ that’s so widely recommended in the JS ecosystem has what feels like a complete coverup. Joins are crucial to the performance of SQL and offload the processing to the database. This is especially important for enterprises running large SQL databases. 

At the very least it should be made clear on the website and in the docs so that others considering switching to Prisma are aware of the consequences of doing so. 

I couldn’t find any reference to this issue in the official docs.

**Reply by lukejagodzinski** [reply]
Yep, it would be best to mention it in the docs. Unfortunatelly, people not knowing SQL created a lot of hype around Prisma which in full fariness is still great. I haven't had better DX experience with any ORM before. Hopefully, they can resolve this issue soon because I don't want to switch.

For now, I'm just considering using Drizzle. I've already done some tests and it's faster but it doesn't have all the features of Prisma yet. They recently introduced very similar API but still is a little bit lacking.

**Reply by janpio** [reply]
We are working on fixing this, Prisma will be using JOINs in more cases. Follow https://github.com/prisma/prisma/issues/5184 for updates.

**Reply by iukea1** [reply]
> Yep, it would be best to mention it in the docs. Unfortunatelly, people not knowing SQL created a lot of hype around Prisma which in full fariness is still great. I haven't had better DX experience with any ORM before. Hopefully, they can resolve this issue soon because I don't want to switch.
> 
> For now, I'm just considering using Drizzle. I've already done some tests and it's faster but it doesn't have all the features of Prisma yet. They recently introduced very similar API but still is a little bit lacking.

I just started using drizzle (I have a backend engineering background and am pretty familiar with SQL

Drizzle is a breath of fresh air and my total query  to page render time is now 12ms  (Versel edge function and did a small table join) 



You only downside..... Is it still extremely new and things are missing ...

 I am currently helping build integration into Next-Auth  (should be out soon)

**Reply by rxliuli** [reply]
I have switched to drizzle, and what shocked me the most is the change in bundle size, which decreased by 90% just by replacing prisma.

<img width="408" alt="image" src="https://github.com/user-attachments/assets/6e9dbf7d-f3c6-4fb5-bc8a-02dffd5ccde2" />
<br/>
<img width="423" alt="image" src="https://github.com/user-attachments/assets/c3c7112f-74ac-4842-b014-8d843eadf397" />

**Reply by coopbri** [reply]
@rxliuli yeah, that is nuts. Drizzle is destroying Prisma in metrics left and right. Competition is good!

**Reply by dy0gu** [reply]
Excuse my ignorance but how is bundle size relevant when it comes to ORMs like Prisma and Drizzle? They run on the server so more or less size will never impact the end user like it would for a package in a web app.

Also, as an aside, it was quite funny getting a notification today from a 2023 thread I am somehow subscribed to.

**Reply by rxliuli** [reply]
> Excuse my ignorance but how is bundle size relevant when it comes to ORMs like Prisma and Drizzle? They run on the server so more or less size will never impact the end user like it would for a package in a web app.

In environments like Cloudflare Workers, package size is one of the factors affecting API performance, as unlike traditional servers that run continuously after initialization, Workers can have cold starts.

> Also, as an aside, it was quite funny getting a notification today from a 2023 thread I am somehow subscribed to.

Because Prisma still has similar problems now... even more problems.

**Comment by kibertoad** [comment]
@janpio Are there plans to ever address that? drizzle managed to implement joins via just one sql query, maybe you could take a look at their approach.

**Reply by iukea1** [reply]
I am waiting to see what Prisma is going to do.

**Reply by janpio** [reply]
We will switch our query generation from separate queries to using JOINs when this is our most pressing concern.

**Reply by kibertoad** [reply]
Why not support both?

**Reply by janpio** [reply]
Maybe we will also support both, but that adds another decision for the user so the default has to be the one that is usually more desirable.

**Reply by kibertoad** [reply]
sure, that makes sense.
are there benchmarks behind current implementation?

**Reply by samulisuomi** [reply]
@janpio please reconsider the arguments mentioned in the other thread why this would be good for performance and keeping the operations ACID:
 https://github.com/prisma/prisma/discussions/12715#discussioncomment-3759280
https://github.com/prisma/prisma/discussions/12715#discussioncomment-5487485
https://github.com/prisma/prisma/discussions/12715#discussioncomment-5631570

I feel like these points haven’t been addressed properly. In the project that I’m starting, I went with Kysely instead of Prisma because of this and other similar limitations.  Is there an intentional philosophy in Prisma to not do many things the ”traditional SQL-way” or why this feature is so quickly dismissed as unnecessary? 🤔

**Reply by janpio** [reply]
@samulisuomi I just wrote that we will work on implementing JOINs. What is there to reconsider?

**Reply by samulisuomi** [reply]
@janpio apologizes if I misunderstood. Earlier you said

> We will switch our query generation from separate queries to using JOINs when this is our most pressing concern.

which to me sounded like this will not be prioritized for quite some time. Do you have any more explicit time frame in mind?

**Reply by janpio** [reply]
No. We unfortunately never communicate timeframes or ETAs. 
That I am responding to you here indicates that this is not at the end of our list though. We have people working on related problems right now.

**Reply by HomiG** [reply]
> @janpio please reconsider the arguments mentioned in the other thread why this would be good for performance and keeping the operations ACID:
>  https://github.com/prisma/prisma/discussions/12715#discussioncomment-3759280
> https://github.com/prisma/prisma/discussions/12715#discussioncomment-5487485
> https://github.com/prisma/prisma/discussions/12715#discussioncomment-5631570
> 
> I feel like these points haven’t been addressed properly. In the project that I’m starting, I went with Kysely instead of Prisma because of this and other similar limitations.  Is there an intentional philosophy in Prisma to not do many things the ”traditional SQL-way” or why this feature is so quickly dismissed as unnecessary? 🤔 

_For anyone who is wondering why Prisma is working this way:_

I think so. To be honest I've reached this issue on GitHub because I was looking at how Prisma is working under the hood and I got surprised. Watching [this](https://youtu.be/RAoXdyI_PH4) conference, you can get a glance at how Prisma has a main core and implements different connectors for each implementation. Either way, towards the end of the video it's explained how a Prisma query is translated. Around +-25 mins it's revealed, that there's a vision to interconnect if I'm not wrong, data from different dbs, even different kind of services (as they say stripe in the video). You'll get a better glance if you see the schema. 
This is the vision, and in order to interconnect data from different resources, from the architectural perspective it's definitely easier to query layer by layer since you can't really join between a db and a service.
Although from the other side if you just have Prisma to query just a single db, it's shocking that there's no joining. That's my opinion after all, and my  understanding of the video.

**Comment by chris-braidwell** [comment]
While it's irritating that Prisma doesn't use joins, there are two distinct issues that should be addressed separately. (1) efficiencies relating to using joins or not, and (2) data correctness relating to using joins or not. (1) is a performance-only issue, important but doesn't affect data integrity directly -- while (2) is an ACID correctness / data integrity issue that's deadly if it's done wrong.

If (2) is done wrong, we don't need to care about (1). Frankly if (2) is done wrong, we should all stop using Prisma immediately.

The key question is, does Prisma wrap **ALL** operations in transactions? If the answer is yes, and we as developers configure our DBs with the right isolation level, then issue (2) is OK. As I read the current Prisma docs, they guarantee transactional behavior for only some operations. See https://www.prisma.io/docs/concepts/components/prisma-client/transactions. That lists: nested writes; bulk transactions; and when the `$transaction` API is used directly. That's not enough. We need to know that absolutely all Prisma operations are transactional. Here are some example scenarios that would be a problem if not:

- Developer calls a `findFirst` method with an `include` clause. If the query (pure read operation) isn't transactional and is broken down by Prisma into multiple SELECTs, then there is a race condition between the SELECTs and the returned data could be different (wrong) compared to the same query if it had been implemented as a single SQL statement (one SELECT with joins).

- Developer calls an `update` method that Prisma breaks down into one UPDATE and one or more SELECTs. Again there is a race condition between the SELECTs and UPDATEs, and a result that's potentially different from what a single UPDATE would provide.

Bottom line: Prisma team, can you assure us that **ALL** operations, meaning all writes and all reads, are transactional? If it's true that all Prisma operations are wrapped in a transaction, there is no potential race condition between multiple INSERTs, UPDATEs and SELECTs generated by Prisma, and we don't have to worry about data integrity problems as a result of issue (2). Then we can get back to griping about the performance aspects in issue (1) :).  (And you should update the docs to say **ALL** is transactional).

However if there are any Prisma operations that aren't transactional (not just writes -- even reads, if Prisma chooses to break the read operation into multiple SELECTs) -- then there is a Severity 1+++++++ data integrity issue. Which seems unlikely but I'd really like to hear the Prisma team say all is transactional, not just the special cases listed on the doco page I referenced above.

**Reply by AnimeshRy** [reply]
Bumping this as I really want to get this before switching out prisma completely

**Reply by AnimaLupi** [reply]
The lack of any kind of reply to this message is really worrying.

**Reply by janpio** [reply]
It's easy not to see comments on random discussions. But here I am:

> The key question is, does Prisma wrap ALL operations in transactions?

No, Prisma does not wrap _all_ operations in transactions.

**Reply by chris-braidwell** [reply]
@janpio that seems like a significant problem. If I wasn't using Prisma and chose to implement an operation as multiple SQL commands, for a safe result I'd need to (1) wrap the commands in a transaction and (2) ensure my session had the right isolation level. When using Prisma, the isolation level is still my responsibility, but the transaction handling should be Prisma's responsibility.

It violates the principle of least amazement for a single Prisma call to be implemented as multiple SQL commands with no transactional protection. I don't think most Prisma users consider this because it isn't made clear in the web site or docs.

Am I missing something?

**Reply by janpio** [reply]
I think I misunderstood your question, or you misunderstood my answer.

Operations that require multiple SQL statements, and that require to be wrapped in transactions for any reason, are of course wrapped in transactions. Operations that do not (so single SQL statement ones), are not.

**Reply by chris-braidwell** [reply]
Glad that I misunderstood. Agreed that's the right behavior.

**Reply by janpio** [reply]
If there are any queries, where we get this even slightly wrong, are bug we would like to see reported and want to fix. So if you find anything (pretty simple to log via https://pris.ly/d/logging), please do that. Very often things are usually fine for many users, but of course in edge cases this can lead to really bad effects - that we want to avoid.

**Reply by nikooiko** [reply]
Hi @janpio, I've created a [test](https://github.com/nikooiko/prisma-transaction-tests) that showcases that prisma doesn't properly handle nested reads and transactions. Based on what you said and the documentation, it would make sense to expect (and if I'm not mistaken I'm not the only one on this) **wrapping a prisma query in a transaction** should result in **all queries generated by prisma run inside that transaction**. _But this is not actually the case_.

If you follow the repo you can clearly see that the 3rd case (mode 'nested-with-transaction' => nested ) fails, which means that prisma doesn't actually follow the convention you mention and that some of the queries are running independently outside the transaction, causing read inconsistencies.

In short (for those bored to read the entire message), **wrapping a prisma query with nested reads in a transaction doesn't mean that all the read operations will run inside this transaction**.

PS (edit): the same is also true for interactive transactions. Nothing changes with that too (check `nested-with-interactive-transaction`)

**Reply by janpio** [reply]
That could be a bug then.

Can you explain a bit more what your script (= the different setups) actually does? 
In the README all Prisma cases seem to be failing, while here you write only about `3rd case (mode 'nested-with-transaction' => nested )`. Is only the third failure a problem or are the others also unexpected?

**Reply by nikooiko** [reply]
> That could be a bug then.
> 
> 
> 
> Can you explain a bit more what your script (= the different setups) actually does? 
> 
> In the README all Prisma cases seem to be failing, while here you write only about `3rd case (mode 'nested-with-transaction' => nested )`. Is only the third failure a problem or are the others also unexpected?

The 2nd case (nested reads) is expected to fail based on what you described above, but the 3rd and the 4th (wrapping nested reads in transaction, interactive or not) aren't. I think that wrapping any query inside a transaction should mean that this and any nested queries should run inside that transaction no matter what, otherwise the transaction doesn't make any sense.

**Reply by janpio** [reply]
Thanks, I am slowly making sense of the script now and ... am confused about what is going on. Will raise this with the team tomorrow.

**Reply by janpio** [reply]
(Temporary explanation/theory: The `$transaction` methods do not trigger as expected when there is only 1 Prisma Client query - which ignores that that is currently executed with 2 independent queries under the hood, and wrapping in a transaction would make so much sense there.)

**Reply by janpio** [reply]
While preparing my internal writeup for tomorrow, I actually noticed a few problems with the current test code. Here is a PR: https://github.com/nikooiko/prisma-transaction-tests/pull/1 With that we get the expected `SUCCESS -> Finished without mismatches for mode 'nested-with-interactive-transaction'` - so I think Prisma is doing what it is expected here.

Also, here is a PR that shows my current theory why the batch transactions don't do their job properly: https://github.com/janpio/prisma-transaction-tests/pull/1 Adding a second query that does nothing triggers the wrapping in a transaction properly, and then correctly also leads to `SUCCESS -> Finished without mismatches for mode 'nested-with-transaction'`. The need for a second query is unexpected and a weird design choice.

To be honest, I am also a bit unsure why the `nested` case fails though, even with my description above. That does not feel very good...

**Reply by nikooiko** [reply]
I think the **nested** case makes sense to fail as prisma runs the nested queries independently. Even if it did wrap them in a transaction (which as you said it doesnt), there would be no way to know what's the appropriate isolation level for it, as in this test the default level one doesn't cover our needs.

**Reply by nikooiko** [reply]
In order for the nested case to be successful, the nested queries should be run as subqueries (or joins instead).

**Reply by janpio** [reply]
Good that we are already working on that 😆 https://github.com/prisma/prisma/issues/5184

**Reply by nikooiko** [reply]
Although using joins will resolve this, IMO it's not 100% related. ie if Prisma never ends up utilizing joins for this (or simply doesn't on all cases), it means that this won't be resolved. I think product-wise it makes sense to tackle this independently to make sure that no matter what this will be resolved at some point. This (at least for me) is very critical as it can easily cause lots of inconsistencies.

**Reply by janpio** [reply]
Usage of a JOIN in your test query would make this problem moot, right? And as we are already working on that (as in: writing code right now), that is the best and most direct solution right now as well I would say.
And for the queries where we still can not use JOINs (and there will be many, as Prisma has a much more flexible API), we want (and have) to use transactions to wrap the individual queries together.

What really confuses me still, and I could not figure out today, is why using `$transaction([...])` does not wrap the query in a transaction. I'll keep investigating.

**Reply by seifsg** [reply]
@janpio any updates regarding this please? Thanks.

**Comment by dy0gu** [comment]
Correct me if I'm wrong but shouldn't this discussion have been closed long ago?

Is the problem mentioned here not the exact case [this feature](https://www.prisma.io/blog/prisma-orm-now-lets-you-choose-the-best-join-strategy-preview) solves? It lets you pick between using joins (now default) or multiple queries.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!

**Comment by useafterfree** [comment]
You can track the join discussion here: https://www.prisma.io/blog/prisma-orm-now-lets-you-choose-the-best-join-strategy-preview
